### PR TITLE
Small speed-ups of sqrt and composite unit creation.

### PR DESCRIPTION
**Pull request by mhvk**
Inspired by #7546, some small speed-ups for units. Some avoiding initialization, some avoiding unnecessary checks, but with some effect:
```
a = np.arange(10.) * u.m
%timeit np.sqrt(a)
# 26.2 -> 18.4 us
```
EDIT: with the second commit, 14.8 us.

### CODE REVIEW

File: astropy/units/core.py [hunk 1]
```diff
@@ -2072,7 +2073,7 @@ def add_unit(unit, power, scale):
         new_parts.sort(key=lambda x: (-x[1], getattr(x[0], 'name', '')))
 
         self._bases = [x[0] for x in new_parts]
-        self._powers = [validate_power(x[1]) for x in new_parts]
```
Comment (mhvk) [inline astropy/units/core.py]: By this time, the power has already been validated, and it cannot change in a way that makes it invalid.
(but scale can change, hence that check remains in place)

File: astropy/units/utils.py [hunk 2]
```diff
@@ -202,7 +202,7 @@ def validate_power(p, support_tuples=False):
     p : float, int, Rational, Fraction
         Power to be converted
     """
-    if isinstance(p, (numbers.Rational, Fraction)):
+    if isinstance(p, (Fraction, numbers.Rational)):
```
Comment (mhvk) [inline astropy/units/utils.py]: We use `Fraction` internally, and `numbers.Rational` is an ABC of which `Fraction` is a member - but `isinstance` on it is much slower. Might be even better to just remove it...

File: astropy/units/utils.py [hunk 3]
```diff
@@ -202,15 +202,8 @@ def validate_power(p, support_tuples=False):
     p : float, int, Rational, Fraction
         Power to be converted
     """
-    if isinstance(p, (numbers.Rational, Fraction)):
```
Comment (mhvk) [inline astropy/units/utils.py]: `isinstance` is slow, especially on an ABC like `Rational`. The duck-typing makes more sense here, and is much faster.

**Comment by mhvk** [comment]
@astrofrog - keeping track of the performance PRs makes sense. But it doesn't really fall within any of the current sections of the change-log. Shall I make a new section titled `Performance Improvements` just for 3.1? (cc @bsipocz)

**Comment by bsipocz** [comment]
Yes, that sounds good. Previously we've mentioned the significant performance improvements at the last section "Other Changes and Additions", but given there'll be multiple ones this time it makes sense to have a separate section for it. 

 Also, we'll need to have a what's new section to summarize all the performance related changes, but I think that will need to be written up using the `Performance` label right before the release.

**Comment by bsipocz** [comment]
Also, a separate question. In the past we've backported some of more straightforward performance improvements to bugfix/LTS. I suspect we don't plan to do that this time around, right?

**Comment by mhvk** [comment]
@bsipocz - I'd be wary of backporting things like this.

On the changelog - I'll add a performance section.

**Comment by pllim** [comment]
Re: Performance section -- Probably a different PR but also need to retroactively include some merged PRs labeled Performance and milestoned by 3.1 (if any).

**Comment by bsipocz** [comment]
@pllim - Good point. So far there were 4 such PRs for 3.1, and only 2 (#7117, and #7087) of those had internal only changes to speed up things, the others had new user facing changes, too. 
Those 2 had their changelog in the new feature section, which probably could work here, too? Is speeding up things a new feature or a bug fix?

**Comment by pllim** [comment]
> Is speeding up things a new feature or a bug fix?

I *think* it is neither? I recently came across a term used for performance improvement task but I don't remember now.

**Comment by mhvk** [comment]
OK, added a changelog entry. (Not sure why circle-ci runs despite the ci-skip).

**Comment by pllim** [comment]
CircleCI doesn't understand `[ci skip]` directive, don't worry about it. :smile:

**Comment by mhvk** [comment]
@adrn - could you have a quick look?

**Comment by adrn** [comment]
Happy to take a look but won't be able to until tomorrow afternoon I'm afraid.

**Comment by adrn** [comment]
I take it back - it wasn't as long as I was expecting so I took a look now :)

This LGTM - thanks @mhvk!

**Comment by mhvk** [comment]
Merged since there were two approvals (and it is good to have the `performance` section ready in the changelog so we can add to it!)

**Pull request merged**